Firestore/core/src/firebase/firestore/core/sync_engine.mm

```diff
@@ -0,0 +1,497 @@
+using model::BatchId;
+using model::DocumentKey;
+using model::DocumentKeySet;
+using model::DocumentMap;
+using model::kBatchIdUnknown;
+using model::ListenSequenceNumber;
+using model::MaybeDocumentMap;
+using model::NoDocument;
+using model::SnapshotVersion;
+using model::TargetId;
+using remote::RemoteEvent;
+using remote::TargetChange;
+using util::AsyncQueue;
+using util::Status;
+using util::StatusCallback;
+
+// Limbo documents don't use persistence, and are eagerly GC'd. So, listens for
+// them don't need real sequence numbers.
+static const ListenSequenceNumber kIrrelevantSequenceNumber = -1;
+
+SyncEngine::SyncEngine(FSTLocalStore* local_store,
+                       remote::RemoteStore* remote_store,
+                       const auth::User initial_user)
+    : local_store_(local_store),
+      remote_store_(remote_store),
+      current_user_(initial_user),
+      target_id_generator_(TargetIdGenerator::SyncEngineTargetIdGenerator()) {
+}
+
+void SyncEngine::AssertCallbackExists(std::string source) {
+  HARD_ASSERT(sync_engine_callback_,
+              "Tried to call '%s' before callback was registered.", source);
+}
+
+TargetId SyncEngine::Listen(Query query) {
+  AssertCallbackExists("Listen");
+  HARD_ASSERT(query_views_by_query_.find(query) == query_views_by_query_.end(),
+              "We already listen to query: %s", query.ToString());
+
+  QueryData queryData = [local_store_ allocateQuery:query];
+  ViewSnapshot viewSnapshot = InitializeViewAndComputeSnapshot(queryData);
+  sync_engine_callback_->OnViewSnapshots({viewSnapshot});
+
+  remote_store_->Listen(queryData);
+  return queryData.target_id();
+}
+
+ViewSnapshot SyncEngine::InitializeViewAndComputeSnapshot(
+    const local::QueryData& query_data) {
+  DocumentMap docs = [local_store_ executeQuery:query_data.query()];
+  DocumentKeySet remote_keys =
+      [local_store_ remoteDocumentKeysForTarget:query_data.target_id()];
+
+  View view(query_data.query(), std::move(remote_keys));
+  ViewDocumentChanges view_doc_changes =
+      view.ComputeDocumentChanges(docs.underlying_map());
+  ViewChange view_change = view.ApplyChanges(view_doc_changes);
+  HARD_ASSERT(view_change.limbo_changes().empty(),
+              "View returned limbo docs before target ack from the server.");
+
+  auto query_view =
+      std::make_shared<QueryView>(query_data.query(), query_data.target_id(),
+                                  query_data.resume_token(), view);
+  query_views_by_query_[query_data.query()] = query_view;
+  query_views_by_target_[query_data.target_id()] = query_view;
+
+  HARD_ASSERT(
+      view_change.snapshot().has_value(),
+      "ApplyChanges to documents for new view should always return a snapshot");
+  return view_change.snapshot().value();
+}
+
+void SyncEngine::StopListening(const Query& query) {
+  AssertCallbackExists("StopListening");
+  auto query_view = query_views_by_query_[query];
+  HARD_ASSERT(query_view, "Trying to stop listening to a query not found");
```

Comment:
It's pretty straightforward to avoid shared ownership here. On the other hand, using a `shared_ptr` avoids having two copies of the same `QueryView` between the two maps, which is probably a good thing. Originally I was going to push towards replacing the `shared_ptr` with a plain value, but on a second thought if might be better to leave as is.

For the current state to work, you're correct that there has to be a copy of the shared pointer in this function.